tests/common/auth_test.go

```diff
@@ -131,3 +140,287 @@ func TestAuthStatus(t *testing.T) {
+			t.Errorf("want error %s but got %v", rpctypes.ErrAuthFailed.Error(), err)
+		}
+	})
+}
+
+func TestAuthRoleRevokeDuringOps(t *testing.T) {
+	testRunner.BeforeTest(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(config.ClusterConfig{ClusterSize: 1}))
+	defer clus.Close()
+	cc := testutils.MustClient(clus.Client())
+	testutils.ExecuteUntil(ctx, t, func() {
+		require.NoError(t, cc.Put(ctx, "foo", "bar", config.PutOptions{}))
+		require.NoErrorf(t, setupAuth(cc, []authRole{testRole}, []authUser{rootUser, testUser}), "failed to enable auth")
+
+		rootAuthClient := testutils.MustClient(clus.Client(WithAuth(rootUserName, rootPassword)))
+		testUserAuthClient := testutils.MustClient(clus.Client(WithAuth(testUserName, testPassword)))
+
+		// create a key
+		require.NoError(t, testUserAuthClient.Put(ctx, "foo", "bar", config.PutOptions{}))
+		// confirm put succeeded
+		resp, err := testUserAuthClient.Get(ctx, "foo", config.GetOptions{})
+		require.NoError(t, err)
+		if len(resp.Kvs) != 1 || string(resp.Kvs[0].Key) != "foo" || string(resp.Kvs[0].Value) != "bar" {
+			t.Fatalf("want key value pair 'foo' 'bar' but got %+v", resp.Kvs)
+		}
+		// create a new role
+		_, err = rootAuthClient.RoleAdd(ctx, "test-role2")
+		require.NoError(t, err)
+		// grant a new key to the new role
+		_, err = rootAuthClient.RoleGrantPermission(ctx, "test-role2", "hoo", "", clientv3.PermissionType(clientv3.PermReadWrite))
+		require.NoError(t, err)
+		// grant the new role to the user
+		_, err = rootAuthClient.UserGrantRole(ctx, testUserName, "test-role2")
+		require.NoError(t, err)
+
+		// try a newly granted key
+		require.NoError(t, testUserAuthClient.Put(ctx, "hoo", "bar", config.PutOptions{}))
+		// confirm put succeeded
+		resp, err = testUserAuthClient.Get(ctx, "hoo", config.GetOptions{})
+		require.NoError(t, err)
+		if len(resp.Kvs) != 1 || string(resp.Kvs[0].Key) != "hoo" || string(resp.Kvs[0].Value) != "bar" {
+			t.Fatalf("want key value pair 'hoo' 'bar' but got %+v", resp.Kvs)
+		}
+		// revoke a role from the user
+		_, err = rootAuthClient.UserRevokeRole(ctx, testUserName, testRoleName)
+		require.NoError(t, err)
+		// check the role is revoked and permission is lost from the user
+		putFailPerm(ctx, testUserAuthClient, "foo", "baz", t)
+
+		// try a key that can be accessed from the remaining role
+		require.NoError(t, testUserAuthClient.Put(ctx, "hoo", "bar2", config.PutOptions{}))
+		// confirm put succeeded
+		resp, err = testUserAuthClient.Get(ctx, "hoo", config.GetOptions{})
+		require.NoError(t, err)
+		if len(resp.Kvs) != 1 || string(resp.Kvs[0].Key) != "hoo" || string(resp.Kvs[0].Value) != "bar2" {
+			t.Fatalf("want key value pair 'hoo' 'bar2' but got %+v", resp.Kvs)
+		}
+	})
+}
+
+func putFailPerm(ctx context.Context, cc interfaces.Client, key, val string, t *testing.T) {
```

Comment:
It's hard to understand what this function does, can you instead rewrite the code:
```
  ...
  putFailPerm(ctx, testUserAuthClient, "foo", "baz", t)
  ...
  
  
func putFailPerm(ctx context.Context, cc interfaces.Client, key, val string, t *testing.T) {
	err := cc.Put(ctx, key, val, config.PutOptions{})
	if err == nil || !strings.Contains(err.Error(), rpctypes.ErrPermissionDenied.Error()) {
		t.Errorf("want error %s but got %v", rpctypes.ErrPermissionDenied.Error(), err)
	}
}
```
to
```
   ...
   err := testUserAuthClient.Put(ctx, "foo", "baz", config.PutOptions{})
   expectPermissionDenied(t, err)
   ...
   
func expectPermissionDenied(t *testing.T, err error) {
    t.Helper()
	if err == nil || !strings.Contains(err.Error(), rpctypes.ErrPermissionDenied.Error()) {
		t.Errorf("want error %s but got %v", rpctypes.ErrPermissionDenied.Error(), err)
	}
}
```

Reply:
Or just 
```
   ...
   err := testUserAuthClient.Put(ctx, "foo", "baz", config.PutOptions{})
   require.ErrorContains(t, err, "permission denied")
   ...
```